File: app/src/main/java/ca/judacribz/gainzassist/activities/add_workout/WorkoutEntry.java
package ca.judacribz.gainzassist.activities.add_workout;

import android.content.Intent;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.Toast;
import butterknife.*;
import ca.judacribz.gainzassist.R;

import static ca.judacribz.gainzassist.constants.ExerciseConst.MIN_INT;
import static ca.judacribz.gainzassist.util.Preferences.getThemePref;
import static ca.judacribz.gainzassist.util.UI.*;

public class WorkoutEntry extends AppCompatActivity{

    // Constants
    // --------------------------------------------------------------------------------------------
    public static final String EXTRA_WORKOUT_NAME
            = "ca.judacribz.gainzassist.activities.add_workout.EXTRA_WORKOUT";
    public static final String EXTRA_NUM_EXERCISES
            = "ca.judacribz.gainzassist.activities.add_workout.EXTRA_NUM_EXERCISES";
    // --------------------------------------------------------------------------------------------

    // Global Vars
    // --------------------------------------------------------------------------------------------
    boolean isEmpty = true;
    int numExs;

    @BindView(R.id.et_workout_name) EditText etWorkoutName;
    @BindView(R.id.et_num_exercises) EditText etNumExercises;
    @BindView(R.id.ibtn_inc_exercises) ImageButton ibtnIncExercises;
    @BindView(R.id.ibtn_dec_exercises) ImageButton ibtnDecExercises;
    @BindView(R.id.btn_enter) Button btnEnter;
    // --------------------------------------------------------------------------------------------

    // AppCompatActivity Override
    ///////////////////////////////////////////////////////////////////////////////////////////////

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setInitView(this, R.layout.activity_workout_entry, R.string.add_workout, true);

        numExs = Integer.valueOf(getString(R.string.initial_num_exercises));
    }

    @Override
    public boolean onSupportNavigateUp() {
        finish();
        return super.onSupportNavigateUp();
    }
    //AppCompatActivity//Override//////////////////////////////////////////////////////////////////


    // TextWatcher Handling
    // =============================================================================================
    @OnTextChanged(value = R.id.et_num_exercises, callback = OnTextChanged.Callback.BEFORE_TEXT_CHANGED)
    public void beforeNumExercisesChanged() {
        setVisibleIfDisabled(ibtnDecExercises);
    }

    @OnTextChanged(value = R.id.et_num_exercises, callback = OnTextChanged.Callback.TEXT_CHANGED)
    public void onNumExercisesChanged(CharSequence s, int start, int before, int count) {
        numExs = handleNumChanged(ibtnDecExercises, s.toString(), MIN_INT).intValue();
    }

    @OnTextChanged(value = R.id.et_workout_name, callback = OnTextChanged.Callback.TEXT_CHANGED)
    public void onWorkoutNameChanged(CharSequence s, int start, int before, int count) {
        if (s.toString().trim().isEmpty()) {
            if (!isEmpty) {
                isEmpty = true;
                btnEnter.setText(R.string.skip);
            }
        } else {
            if (isEmpty) {
                isEmpty = false;
                btnEnter.setText(R.string.enter);
            }
        }
    }
    // =TextWatcher=Handling========================================================================


    // OnFocusChanged Handling
    // =============================================================================================
    @OnFocusChange(R.id.et_num_exercises)
    void onFocusLeft(EditText et, boolean hasFocus) {
        if (!hasFocus) {
            handleFocusLeft(et, MIN_INT, numExs);
        }
    }
    // =OnFocusChanged=Handling=====================================================================

    // Click Handling
    // =============================================================================================
    /* Increase number of num_sets */
    @OnClick(R.id.ibtn_inc_exercises)
    public void incNumExs() {
        setText(etNumExercises, numExs + 1);
    }

    /* Decrease number of num_sets */
    @OnClick(R.id.ibtn_dec_exercises)
    public void decNumExs() {
        setText(etNumExercises, numExs - 1);
    }

    @OnClick(R.id.btn_enter)
    public void enterWorkoutName() {
        if (validateFormEntry(this, etNumExercises)) {
            Intent exercisesEntry = new Intent(this, ExercisesEntry.class);

            if (!isEmpty) {
                Toast.makeText(this, "" + getTextString(etWorkoutName), Toast.LENGTH_SHORT).show();
                exercisesEntry.putExtra(EXTRA_WORKOUT_NAME, getTextString(etWorkoutName));
            }


            exercisesEntry.putExtra(EXTRA_NUM_EXERCISES, Math.max(MIN_INT, getTextInt(etNumExercises)));
            startActivity(exercisesEntry);
        }
    }

    @OnClick(R.id.btn_cancel)
    public void cancelWorkout() {
        finish();
    }
    //=Click=Handling===============================================================================

}

File: app/src/main/java/ca/judacribz/gainzassist/activities/add_workout/WorkoutEntry.kt
package ca.judacribz.gainzassist.activities.add_workout

import android.content.Intent
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.widget.EditText
import android.widget.Toast
import ca.judacribz.gainzassist.R
import ca.judacribz.gainzassist.constants.ExerciseConst.MIN_INT
import ca.judacribz.gainzassist.databinding.ActivityWorkoutEntryBinding
import ca.judacribz.gainzassist.util.UI.setInitTheme
import ca.judacribz.gainzassist.util.UI.setToolbar
import ca.judacribz.gainzassist.util.UI.setVisibleIfDisabled
import ca.judacribz.gainzassist.util.UI.handleNumChanged
import ca.judacribz.gainzassist.util.UI.handleFocusLeft
import ca.judacribz.gainzassist.util.UI.setText
import ca.judacribz.gainzassist.util.UI.validateFormEntry
import ca.judacribz.gainzassist.util.UI.getTextString
import ca.judacribz.gainzassist.util.UI.getTextInt

class WorkoutEntry : AppCompatActivity() {

    companion object {
        const val REQ_EXERCISES_ENTRY = 1001
        const val EXTRA_WORKOUT_NAME = "ca.judacribz.gainzassist.activities.add_workout.EXTRA_WORKOUT"
        const val EXTRA_NUM_EXERCISES = "ca.judacribz.gainzassist.activities.add_workout.EXTRA_NUM_EXERCISES"
    }

    var isEmpty = true
    var numExs: Int = 0

    private lateinit var binding: ActivityWorkoutEntryBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setInitTheme(this)
        binding = ActivityWorkoutEntryBinding.inflate(layoutInflater)
        setContentView(binding.root)
        setToolbar(this, R.string.add_workout, true)

        numExs = getString(R.string.initial_num_exercises).toInt()

        setupListeners()
    }

    private fun setupListeners() {
        binding.partNumExercises.etNumExercises.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
                setVisibleIfDisabled(binding.ibtnDecExercises)
            }

            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                numExs = handleNumChanged(binding.ibtnDecExercises, s.toString(), MIN_INT).toInt()
            }

            override fun afterTextChanged(s: Editable?) {}
        })

        binding.partWorkoutName.etWorkoutName.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}

            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                if (s.toString().trim { it <= ' ' }.isEmpty()) {
                    if (!isEmpty) {
                        isEmpty = true
                        binding.btnEnter.setText(R.string.skip)
                    }
                } else {
                    if (isEmpty) {
                        isEmpty = false
                        binding.btnEnter.setText(R.string.enter)
                    }
                }
            }

            override fun afterTextChanged(s: Editable?) {}
        })

        binding.partNumExercises.etNumExercises.onFocusChangeListener = View.OnFocusChangeListener { v, hasFocus ->
            if (!hasFocus && v is EditText) {
                handleFocusLeft(v, MIN_INT, numExs)
            }
        }

        binding.ibtnIncExercises.setOnClickListener { incNumExs() }
        binding.ibtnDecExercises.setOnClickListener { decNumExs() }
        binding.btnEnter.setOnClickListener { enterWorkoutName() }
        binding.btnCancel.setOnClickListener { cancelWorkout() }
    }

    override fun onSupportNavigateUp(): Boolean {
        finish()
        return super.onSupportNavigateUp()
    }

    override fun onActivityResult(req: Int, res: Int, data: Intent?) {
        if (res == RESULT_OK) {
            finish()
        }
    }

    fun incNumExs() {
        setText(binding.partNumExercises.etNumExercises, numExs + 1)
    }

    fun decNumExs() {
        setText(binding.partNumExercises.etNumExercises, numExs - 1)
    }

    fun enterWorkoutName() {
        if (validateFormEntry(this, binding.partNumExercises.etNumExercises)) {
            val exercisesEntry = Intent(this, ExercisesEntry::class.java)

            if (!isEmpty) {
                Toast.makeText(this, "" + getTextString(binding.partWorkoutName.etWorkoutName), Toast.LENGTH_SHORT).show()
                exercisesEntry.putExtra(EXTRA_WORKOUT_NAME, getTextString(binding.partWorkoutName.etWorkoutName))
            }

            exercisesEntry.putExtra(EXTRA_NUM_EXERCISES, Math.max(MIN_INT, getTextInt(binding.partNumExercises.etNumExercises)))
            startActivityForResult(exercisesEntry, REQ_EXERCISES_ENTRY)
        }
    }

    fun cancelWorkout() {
        finish()
    }
}
